WIP: Quaternion

ThomasD13 opened the pull request:
### Purpose of this PR

I suspect that at least the given euler angles of this ctor
`        /// <summary>
        /// Construct a new Quaternion from given Euler angles
        /// </summary>
        /// <param name="pitch">The pitch (attitude), rotation around X axis</param>
        /// <param name="yaw">The yaw (heading), rotation around Y axis</param>
        /// <param name="roll">The roll (bank), rotation around Z axis</param>
        public Quaternion(float pitch, float yaw, float roll)`
don't result in a correct Quaternion representation.

1. Step: Adding test project which covers functionality of class Quaternion.
2. Step: Reveal potential bugs of Quaternion.
3. Step: Provide potential bugfix
* Which part of OpenTK does this affect (Math, OpenGL, Platform, Input, etc). **OpenTK.Math.Quaternion**

### Testing status

Still in progress. Covering API of Quaternion with unit tests.
To run these tests just execute "build.sh RunTests"

### Comments

This is still in progress. I created this early push request to get early comments/suggestion/help.
However, this was my very first contact with F# and fake. I managed to get my test project built and executed with NUnit with the provided build.sh script. 
Unfortunately, this isn't pretty (hardcoded). Maybe someone could provide an improvement to distinct between xUnit and NUnit projects.

I've chose NUnit (+Csharp), because I am most familiar with it. Maybe someone (or me, when more time available) could use test criterias as base and move it to xUnit.

Frassle commented on build.fsx:
```diff
@@ -163,6 +164,15 @@ Target "RunTests" (fun _ ->
             XmlOutputPath = Some "TestResults.xml" })
 )
 
+// --------------------------------------------------------------------------------------
```
nunit instead of xunit does make this a little messier to handle, if this was xunit it would of just been picked up by RunTests

Nihlus commented:
Hey! Thanks a ton for your contribution. Unfortunately, I'm not keen on mixing unit testing frameworks, and we really won't be able to merge this as it is. We've been discussing internally about moving the existing unit tests over to C#, and I think this PR speaks volumes for the benefits of doing so.

ThomasD13 commented:
Hi, thanks for your feedback. I understand your arguments without an doubt.
I just want to improve the code, and give you the possibility to verify this with reproducable and automated tests. Thus, things got messy.
However, starting discussion about establishing a standart (using C# for unit tests) is great to hear for me.
The question is how to proceed from here?
I could move forward and get steps 1,2 and 3 done and push this into this PR.
You could criticise the bugfix/testcases, so I get a "ready to merge" codebase.
Meanwhile, if someone take an decision about used language for unit tests, I can refresh my knowledge about xUnit and convert my nUnit tests + take the reviewed bugfix into a new PR.
Or if F# will be the result, someone else could pick up the unit tests and convert them.

Frassle commented:
I'm working on adding a C# xUnit test project to the main develop branch. Once that's in it should be fairly easy for you to rebase your changes on top of that.

ThomasD13 commented:
@Frassle, I used #708 as reference for the paket.exe stuff and csproj.
Compared to current development branch, I added a new project "tests/OpenTK.Tests.CSharp".
This could be a kind of counterpart to "tests/OpenTK.Tests", apart the used language is C# not F#.
I don't know if my nameing is good, since you would distinct between real unit tests, integration or system tests.

Now at this state, code of `Quaternion.cs` improved, but I am not finished yet. I think I have found the wrong code and all affected API, but I can not give you 100% guarantee yet that everything works as expected (proven by testcode). Still open points (at least to me)

1.  Verify the applied rotation order. First should be xAxis, second yAxis, third zAxis. Cover this with unit tests. There could be potential an issue still, but I don't think that.

2. Verify the counter clockwise rotation, given by the right hand rule. That means, if you view to the positive axis, the rotation is applied counter clockwise. 

3. Improve unit test code: After applying some "rotation"-math, it is not easy to verify the result against the expected result in terms of precision loss. I added a helper method which helps  me in a crumble way to determine if a rotation about a **single** axis was calculated correctly. I'm not happy about this. Maybe it worth to get something like `MathHelper.ApproximatelyEqual`into the game, but I don't like this either. It works against the idea of **unit** testing.
However, I will need something like this, at least when I will start to test the rotation math with multiple rotation axis. Any ideas are welcome!

4. Check also class 'Quaterniond'. I haven't had a glance yet, if this suffers under the same problems.

Frassle commented on tests/OpenTK.Tests.CSharp/Quaternion_Tests.cs:
```diff
@@ -0,0 +1,215 @@
+		/// or does not contain 0 when it should be</returns>
+		/// <param name="toTest">To test</param>
+		/// <param name="expected">Expected directions. Values getting only 0 checked</param>
+		public static bool VerifyEuqalSignleDirection(Vector3 toTest, Vector3 expected)
```
Euqal -> Equal
Signle -> Sign

drdbkarron commented:
Can you post known good test results? I reported a quaternion bug a year ago. I had a persistent wobble in quat rotations and abandoned this package. I will revalidate it shortly. I can give you the test code that displayed a wobble rotating a test arrow about its own axis.  If the code is working, the arrow should
not display a wobble at any heading and rotate smoothly about its own local axis.

ThomasD13 commented:
Cool! I appreciate any test data which led to awkward effects. However, I'm still away from that kind of testing level. I just (hopefully) corrected basic stuff, but will extend further and further the testbase.
But I will have a look into that. Are you sure that you didn't suffer under gimbal-lock effect? Sounds like it.

Thanks in advance!
